### ISSUE TITLE
Visual artifacts can appear in the WKWebView

### ISSUE THREAD

**Issue by vincode-io** [Ranchero-Software/NetNewsWire#1911]
I suspect we will need to rotate in the WKWebViews like on iOS to prevent our JavaScript rendering code from causing problems.
![Screen Shot 2020-03-14 at 5 38 50 PM](https://user-images.githubusercontent.com/16448027/76691669-42dab100-661b-11ea-8edd-cb87b13f0bb9.png)

**Comment by vincode-io** [comment]
I was doing the analysis on how to share the WKWebView recycling and rotating code from iOS to macOS and started having flashbacks.  All the problems we had with rendering bugs, blank views, text sizing bugs, etc...  I wondered if it was worth the hassle.

I don't think it is.  I think that the NNW 5.0.x series loads articles fast enough with `loadHTMLString` and that it is hard to spot the performance difference between it and the 5.1 JavaScript loading code.  I'm even doubting myself with regards to the iOS code and wonder if I wasn't doing something wrong that caused the `loadHTMLString` code to run so much slower than the JavaScript loading code.  Maybe we need to revisit decision on iOS if we keep seeing bugs.

I can go ahead and make the iOS code sharable between it and macOS, but my recommendation is to revert to NNW 5.0.x article rendering.

**Comment by vincode-io** [comment]
The downside to reverting to using `loadHTMLString` is that less article rendering code is shared between macOS and iOS.  This could be a bit of a hassle and not something I'm looking forward to.

**Comment by brentsimmons** [comment]
Let’s stick with loadHTMLString on Mac. The performance has been just fine — I’ve had zero problems with it, and I’m a stickler for performance. (And I’m using NNW mostly on an old 2013 MacBook Air.)

**Comment by brentsimmons** [comment]
Is there a way we could test loadHTMLString performance on iOS? If we could go back to that, that would be cool, just since it’s so simple, and we’d have the same code on Mac and iOS.

**Comment by vincode-io** [comment]
Let me think about a way we can test that.  I'm thinking the same thing.  

We have a lot of complexity tied up in making the JavaScript rendering work.  I'd prefer to use `loadHTMLString`, but last time I tried it it was noticeably slower.  Maybe something else was factored in that I missed that was causing it to be so slow.

**Comment by vincode-io** [comment]
I got a version of `loadHTMLString` working for Mac 5.1 and performance is the same as Mac 5.0.x.  I was even able to keep the ArticleRenderer and other shared iOS code intact.

The first few tries weren't good.  `loadHTMLString` was noticeably slower.  I eventually figured out that I was incorrectly referencing our local JavaScript files and WKWebKit was trying to download them from the web.  Fixing that, fixed performance.

I'm now wondering if I didn't have the same problem when I was testing `loadHTMLString` on iOS.  I'd like to implement `loadHTMLString` on iOS and see how fast it is.  I'm not sure I'm comfortable doing it in the ios-release branch.  A longer testing cycle might be good because I'm going to have to unwind some stuff I put in to deal with the JavaScript code.  This might be something better left to the iOS 5.1 release.

@brentsimmons, do you think we should try it for iOS 5.0.1 or wait until iOS 5.1?

**Comment by brentsimmons** [comment]
Let’s try it for iOS 5.0.2. I’d rather get 5.0.1 done soonest.

**Comment by vincode-io** [comment]
If we’re going to try to get it in the 5.0.x series, I’d rather we do it in 5.0.1.  I could have it done by tomorrow evening and ready to test.

**Comment by brentsimmons** [comment]
Okay — agreed.

**Comment by vincode-io** [comment]
I did a bunch of speed comparisons for the JavaScript loading vs. `loadHTMLString`.  I dug in using Safari's timeline and think I have a handle on the differences.  It all comes down to how fast your network connection is.

`loadHTMLString` is slower the slower your network is.  That's why my testing in the past showed `loadHTMLString` to be so much slower than the JavaScript stuff on iOS.  My phone was on an WiFi connection.  It's also why I couldn't tell much difference between the two on my Mac.  It's on a gigabit ethernet connection.  If you completely disconnect your computer from the network, `loadHTMLString` seems to be every bit as fast as the JavaScript rendering.

Looking at Safari's timeline tool, it looks like `loadHTMLString` will attempt to download image assets before rendering sometimes.  Probably because image size info wasn't included in the HTML and it wants it to make rendering the page look less jumpy.

The JavaScript rendering seems to do the opposite with images.  It renders the DOM with the missing image data and then does layout again as the images come in.

Assuming I'm right, we have to decide what we want for article rendering.  Slower, less jumpy or snappy with some jumpiness as image assets come in.

I guess there are other things to take into consideration.  Our image resizing bug might not have existed if we had used `loadHTMLString` on iOS.  Complexity also goes up on macOS if we go with JavaScript because we have to start rotating web views to avoid rendering issues.  On iOS we will still do some rotating because we are constantly creating WebViewControllers and need to avoid the white flashing bug, so not much difference between the two there.
